src/keyszer/lib/key_context.py

```diff
@@ -2,9 +2,21 @@
 
 
 class KeyContext:
-    def __init__(self, device_name):
-        self._device_name = device_name
+    def __init__(self, device):
         self._X_ctx = None
+        leds_list = []
+
+        # Must declare these here or app will crash if keyboard device hasn't been "grabbed" yet
+        self._capslock_state = ""
+        self._numlock_state = ""
+
+        # Check for actual device name being present before using evdev's ".leds()" method
+        # or device.name (doesn't like strings)
+        if not device == "":
```

Comment:
Why would device ever be an empty string?

Reply:
Don’t know why, but if you print out “device” from various locations you’ll notice that it is empty at startup, and is filled as soon as you press a key on the keyboard. `on_event` is a good place, but you have to print it as “{event, device}” or it gives a value error. Not sure if it’s actually ever a string anymore since I modified the PR. I haven’t tried removing that line to see if it will still work without crashing.

Reply:
Tried taking the `if` line out, it definitely still crashes complaining about `device.name` or `device.leds()` not being valid for a string. So even though I'm no longer ever explicitly creating an empty string with an optional pre-defined parameter, device is still an empty string prior to the first physical key press on the keyboard.

Reply:
You need to learn how to read the stacktrace - Python will tell you *exactly which line the problem is on* it's right there mixed in with the error message.  It's possible at boot-up when we're sending fake keys there is no device, but this should be easy to determine from the stacktrace.

Reply:
Oh, I do get a significant amount of help from the stacktrace error. However it usually refers to several different files, and I'm still not very good at quickly working through the whole stack to understand what's really going wrong. I hear that error messages are actually going to be clearer in Python 3.11, possibly even with suggestions about solutions. 

I think that part where the code sends Left_Shift at startup to force `xmodmap` to respect/retain the new keyboard mapping is definitely part of this. But I only have a vague idea what the "correct" solution to that would be. The `device.name` probably won't mind a completely fake device string, but I wonder if `device.leds()` will work with that or still throw an error. I'll try something like that when I have a bit more time.

Reply:
> I think that part where the code sends Left_Shift at startup to force xmodmap to respect/retain the new keyboard mapping

That would also be my guess... :)  I'll try to dig into this later.  We either need a conditional here (but not comparing ""), or we need to change how that startup code works perhaps to directly interface with output.

Reply:
Well, you could knock me over with a feather. Looks like moving everything out of `init` was the key to everything. Check the new form of `KeyContext`. Works without any `if` test. No crash at startup. Everything still seems to respond as desired. 

Testing one last issue... Wow, even the out-of-sync problem when a lock key is enabled prior to startup seems to be resolved. I would hesitantly say this might actually be ready for prime time. Hope you have a numpad handy for testing.

Reply:
> Looks like moving everything out of init was the key to everything.

Yeah. :-) This is why you don't do a bunch of work in `init`. :-)  It only crashes if you try and access device when there is none... so don't access device... the init code that pushes and releases shift doesn't query device - so it won't crash.